acsys/defaults/config.py

```diff
@@ -0,0 +1,220 @@
+        },
+        "test": {
+            "proxy_hostname": "test.fnal.gov",
+            "proxy_port": 80,
+            "request_timeout_ms": 2000,
+            "async_timeout_ms": 1000
+        }
+    }
+
+    """ A list of supported output types to return the contents of the current 
+        configuration state
+    """
+    dump_formats = ['json']
+
+    def __post_init__(self, default_environment, additional_config={}):
+        """ Initializes the configuration object with defaults
+
+        If environment is not specified in the init arguments, the environment
+        will default to "production".
+        """
+        self.hostname = ""
+        self.port = 0
+        self.request_timeout = 0
+        self.async_timeout = 0
+        self.environment = default_environment
+
+        [setattr(self, k, v) for k, v in self.additional_config.items()]
+
+    def dump(self, format="json"):
+        """ Outputs the current state of the configuration in the format specified
+ 
+        The default format is JSON.  The method will check for valid format types.
+        NB: Currently the ONLY supported output format is JSON
+        """
+        output = ""
+        output_format = format.lower()
+        
+        if not output_format in Defaults.dump_formats: 
+            raise ValueError("Argument value for format, '{format}', is not a valid value".format(format=format))
+
+        if output_format == "json":
+            template =  "{{ '{environment}': {{ 'proxy_hostname': '{hostname}', 'proxy_port': {port}, " \
+                        "'request_timeout_ms': {request_timeout}, 'async_timeout_ms': {async_timeout} }} }}"
+            output = template.format(environment=self.environment, hostname=self.hostname, port=self.port, \
+                                     request_timeout=self.request_timeout, async_timeout=self.async_timeout)
+        
+        return output
+
+    @property
+    def environment(self):
+        return self._environment
+
+    @environment.setter
+    def environment(self, environment):
+        """ Sets the environment property.
+        
+        Switching environments will overwrite all config values with the 
+        defaults for the selected environment.  Items are exposed as properties
+        and can be individually overridden using the property syntax.
+        """
+        valid_environments = Defaults.default_enumeration.keys()
+        if environment in valid_environments:
+            self._environment = environment
+            self.hostname = Defaults.default_enumeration[environment]["proxy_hostname"]
+            self.port = Defaults.default_enumeration[environment]["proxy_port"]
+            self.request_timeout =  Defaults.default_enumeration[environment]["request_timeout_ms"]
+            self.async_timeout =  Defaults.default_enumeration[environment]["async_timeout_ms"]
+        else:
+            raise ValueError("Environment '{env}' is not a valid environment".format(env=environment))
```

Comment:
How about:
```suggestion
    try:
        def_env = Defaults.default_enumeration[environment]
        self._environment = environment
        self.hostname = def_env["proxy_hostname"]
        self.port = def_env["proxy_port"]
        self.request_timeout =  def_env["request_timeout_ms"]
        self.async_timeout =  def_env["async_timeout_ms"]
    except KeyError:
        raise ValueError("Environment '{env}' is not a valid environment".format(env=environment))
```

Instead of testing to see if the look-up would work, do the lookup to test its existence *and* to get the default values for that config in one line. The `NotFound` exception can be converted to your more-informative exception.

Reply:
I like Rich's solution!

Reply:
:face_palm: yes, I should have read this before.